monai/data/grid_dataset.py

```diff
@@ -193,25 +210,160 @@ def __init__(
+            indices = list(range(self.cache_num))
+        if self.progress and not has_tqdm:
+            warnings.warn("tqdm is not installed, will not show the caching progress bar.")
+        with ThreadPool(self.num_workers) as p:
+            if self.progress and has_tqdm:
+                return list(tqdm(p.imap(self._load_cache_item, indices), total=len(indices), desc="Loading dataset"))
+            return list(p.imap(self._load_cache_item, indices))
+
+    def _load_cache_item(self, idx: int):
+        """
+        Args:
+            idx: the index of the input data sequence.
+        """
+        item = self.data[idx]  # type: ignore
+        patch_cache, other_cache = [], []
+        for patch, *others in self.patch_iter(item):
+            if self.patch_transform is not None:
+                first_random = self.patch_transform.get_index_of_first(
+                    lambda t: isinstance(t, RandomizableTrait) or not isinstance(t, Transform)
+                )
+                patch = self.patch_transform(patch, end=first_random, threading=True)
+
+            if self.as_contiguous:
+                patch = convert_to_contiguous(patch, memory_format=torch.contiguous_format)
+            if self.with_coordinates and len(others) > 0:  # patch_iter to yield at least 2 items: patch, coords
+                other_cache.append(others[0])
+            patch_cache.append(patch)
+        self._cache_other.append(other_cache)
+        return patch_cache
+
+    def __iter__(self):
+        if self.cache:
+            cache_index = None
+            for image in super().__iter__():
+                key = self.hash_func(image)
+                if key in self._hash_keys:
+                    # if existing in cache, try to get the index in cache
+                    cache_index = self._hash_keys.index(key)
+                if cache_index is None:
+                    # no cache for this index, execute all the transforms directly
+                    for patch, *others in self.patch_iter(image):
+                        out_patch = patch
+                        if self.patch_transform is not None:
+                            out_patch = apply_transform(self.patch_transform, patch, map_items=False)
+                        if (
+                            self.with_coordinates and len(others) > 0
+                        ):  # patch_iter to yield at least 2 items: patch, coords
+                            yield out_patch, others[0]
+                        else:
+                            yield out_patch
+
+                if self._cache is None:
+                    raise RuntimeError("cache buffer is not initialized, please call `set_data()` first.")
+                data = self._cache[cache_index]  # type: ignore
+                other = self._cache_other[cache_index]  # type: ignore
+
+                # load data from cache and execute from the first random transform
+                if not isinstance(self.patch_transform, Compose):
```

Comment:
This check should be in the constructor?

Reply:
Yes. I just followed the CacheDataset to add this check here, perhaps we can remove this check since `self.patch_transform` has been hard-coded to be Compose in the constructor. What do you think?
https://github.com/Project-MONAI/MONAI/blob/cf40cbc833aa076706d9ecacae3eb4e20736f66a/monai/data/grid_dataset.py#L224

Reply:
It can go then yes if we're sure the value of `patch_transform` is expected to change ever.